### PR TITLE
Feature: "SFDX: Open Documentation" (on web)

### PR DESCRIPTION
**Pull request by jeffb-sfdc**
### What does this PR do?
New feature to open up our docs

### What issues does this PR fix or reference?
#https://github.com/forcedotcom/easywriter/issues/85, @W-11683184, W-1168318r@

### CODE REVIEW

File: packages/salesforcedx-apex-debugger/src/messages/i18n.ja.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@
  * decorations, e.g., $(x) uses the https://octicons.github.com/ and should not
  * be localized
  *
- * If ommitted, we will assume _message.
+ * If omitted, we will assume _message.
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-apex-debugger/src/messages/i18n.ja.ts:16]: spelling fix

File: packages/salesforcedx-vscode-apex-replay-debugger/src/messages/i18n.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ export const messages = {
   up_to_five_checkpoints:
     'You have %s of the maximum 5 enabled checkpoints. Remove or disable checkpoints until 5 or fewer are active.',
   no_enabled_checkpoints:
-    'You don\'t have any checkpoints enabled. Enable a checkpoint and try again.',
+    "You don't have any checkpoints enabled. Enable a checkpoint and try again.",
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-apex-replay-debugger/src/messages/i18n.ts:25]: This change was auto-made when I saved.

File: packages/salesforcedx-vscode-core/package.nls.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
   "force_auth_access_token_authorize_org_text": "SFDX: Authorize an Org using Session ID",
   "force_auth_logout_all_text": "SFDX: Log Out from All Authorized Orgs",
   "force_auth_logout_default_text": "SFDX: Log Out from Default Org",
+  "force_open_documentation_text": "SFDX: Open Documentation",
```
Comment (jeffb-sfdc) [inline packages/salesforcedx-vscode-core/package.nls.json:8]: I spoke with @sbudhirajadoc and she would like the command to be, "SFDX: Open Documentation"

File: packages/salesforcedx-apex-debugger/src/messages/i18n.ja.ts (hunk 1 above)
Comment (gbockus-sf) [inline packages/salesforcedx-apex-debugger/src/messages/i18n.ja.ts:16]: ha! that explains the 26 files...I was like 👀

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  'https://developer.salesforce.com/tools/vscode/en/functions/overview';
+export const defaultDocUrl = 'https://developer.salesforce.com/tools/vscode';
+
+const auraPath = '/force-app/main/default/aura/';
+const apexClassesPath = '/force-app/main/default/classes/';
+const lwcPath = '/force-app/main/default/lwc/';
+const functionsPath = '/functions/';
+
+const apexExtension = '.apex';
+const soqlExtension = '.soql';
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:29]: The const above seem like they are reusable outside this command (esp the path/extension values).  Think we could move these all to `packages/salesforcedx-utils-vscode/src/constants.ts`?

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    } else if (filePath.includes(lwcPath)) {
+      docUrl = lwcDocUrl;
+    } else if (filePath.includes(functionsPath)) {
+      docUrl = functionsDocUrl;
+    }
+  }
+
+  if (docUrl === '') {
+    docUrl = defaultDocUrl;
+  }
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:56]: nice job handling the case where it doesn't get an active editor.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  soqlDocUrl
+} from '../../../src/commands';
+
+describe('forceOpenDocumentation', () => {
+  let sb: SinonSandbox;
+  let activeTextEditorStub: SinonStub;
+  let openExternalStub: SinonStub;
+
+  beforeEach(() => {
+    sb = createSandbox();
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:27]: This should go outside the before each I think https://sinonjs.org/releases/latest/sandbox/

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        fileName: '/force-app/main/default/staticresources/example-image.png'
+      }
+    }));
+
+    await forceOpenDocumentation();
+
+    expect(openExternalStub.getCall(0).args[0].toString()).to.equal(
+      defaultDocUrl
+    );
+  });
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:124]: nice tests.  Are there any failure conditions we can verify? What happens if `vscode.env.openExternal` throws?

**Review by gbockus-sf — COMMENTED** [review]
a few comments to consider.  I can QE after you check em out.

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+export const auraDocUrl =
+  'https://developer.salesforce.com/tools/vscode/en/aura/writing';
+export const apexDocUrl =
+  'https://developer.salesforce.com/tools/vscode/en/apex/writing';
+export const soqlDocUrl =
+  'https://developer.salesforce.com/tools/vscode/en/soql/soql-builder';
+export const lwcDocUrl =
+  'https://developer.salesforce.com/tools/vscode/en/lwc/writing';
+export const functionsDocUrl =
+  'https://developer.salesforce.com/tools/vscode/en/functions/overview';
```
[the remark this answers is not in the document]
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:20]: @sbedre are there localized versions of these pages?

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts (hunk 6 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:27]: That's not what we have in our code.  Most(and I thought all) of our test code have the createSandbox() call in the beforeEach().  Searching the codebase, I found...
packages/salesforcedx-utils-vscode/test/unit/commands/channelService.test.ts
packages/salesforcedx-utils-vscode/test/unit/commands/commandletExecutor.test.ts
packages/salesforcedx-utils-vscode/test/unit/helpers/traceFlags.test.ts
.
.
.

There are some with it outside the beforeEach(), like in packages/salesforcedx-utils-vscode/test/unit/helpers/paths.test.ts, but the majority of our code has it in there.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts (hunk 7 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:124]: I wasn't able to think of and failure conditions, and I haven't found cases where openExternal fails.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts (hunk 6 above)
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:27]: Yup, I think it's wrong in most of our code :)

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:20]: I think only ja and en are supported looking at the site (basically en or ja in the URL). @sbudhirajadoc please correctly me if I'm wrong.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts (hunk 7 above)
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:124]: It's a corner case, but what if the user doesn't have a default browser? It's usually good to know the behaviour (via a unit test) even if we don't expect failure. It documents what we expect to happen.  Like if we can't open a browser we should probably be outputting to the user that something went wrong verse just silently doing nothing.

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:20]: done

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts (hunk 4 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:29]: It turns out there are 11 files named, "constants.ts". There doesn't seem to be any occurrences of constants being using from other packages (like utils) so I placed the extension constants in packages/salesforcedx-vscode-core/src/constants.ts,   Let me know if you think they should be in a different file.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts (hunk 7 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:131]: This is something we get automatically from VSC, and I think writing a unit test for it is writing a test validating VS Code, and doesn't validate our code.  As an example, I ran this:
```
export async function forceOpenDocumentation() {
  let docUrl = '';
  const editor = vscode.window.activeTextEditor;
  if (editor) {
    const filePath = editor.document.fileName;
    const extension = path.extname(filePath);

    throw Error('OK dont do that');
.
.
.
```

And the user sees this:
image

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts (hunk 6 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts]: done

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts (hunk 7 above)
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceOpenDocumentation.test.ts:131]: cool.  The main thing I'd have expected in the unit test would be validating that the error thrown by the called function is what is get thrown from our calling code.  Granted there's not error handling here so we don't have anything extra to verify so I'm cool with just calling it a day. Error conditions is just always something to consider in unit tests.

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts (hunk 4 above)
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts]: 👍 that works.  we can figure out the right place for all the constants later.

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import {
+  APEX_FILE_NAME_EXTENSION,
+  SOQL_FILE_NAME_EXTENSION
+} from '../constants';
+import { nls } from '../messages';
+
+const auraPath = '/force-app/main/default/aura/';
+const apexClassesPath = '/force-app/main/default/classes/';
+const lwcPath = '/force-app/main/default/lwc/';
+const functionsPath = '/functions/';
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:19]: nit: FWIW I would have also put these in the constants file.  Same deal where any time we need the auraPath we should always use the same value across the codebase.  First other place it could be used is in the unit tests to build the stub responses.

**Review by gbockus-sf — COMMENTED** [review]
Thanks for the updates...started QE

**Review by gbockus-sf — APPROVED** [review]
Command run: SFDX: Open Documentation

* ✅ Open Apex section from Apex code
* ✅ Aura from an Aura Component
* ✅ Main site from a random non-SF file
* ✅ Soql doc from soql builder
* ✅ Functions
* ✅ Lightning

File: packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts (hunk 9 above)
Comment (jeffb-sfdc) [reply packages/salesforcedx-vscode-core/src/commands/forceOpenDocumentation.ts:19]: I'll go and move them
